storage/rocksdb/rdb_io_watchdog.cc

```diff
@@ -0,0 +1,233 @@
+  // `m_write_timeout` seconds. We'll log a message and shut down the service.
+  // NO_LINT_DEBUG
+  sql_print_error("MyRocks has detected a combination of I/O requests which "
+                  "have cumulatively been blocking for more than %u seconds. "
+                  "Shutting the service down.",
+                  m_write_timeout);
+
+  abort_with_stack_traces();
+}
+
+void Rdb_io_watchdog::io_check_callback(union sigval timer_data) {
+  RDB_MUTEX_LOCK_CHECK(m_reset_mutex);
+
+  DBUG_ASSERT(timer_data.sival_ptr != nullptr);
+
+  struct sigevent e;
+
+  e.sigev_notify = SIGEV_THREAD;
+  e.sigev_notify_function = &Rdb_io_watchdog::expire_io_callback_wrapper;
+  e.sigev_value.sival_ptr = this;
+  e.sigev_notify_attributes = nullptr;
+
+  int ret = timer_create(CLOCK_MONOTONIC, &e, &m_io_check_watchdog_timer);
+
+  if (unlikely(ret)) {
+    // NO_LINT_DEBUG
+    sql_print_warning("Creating a watchdog I/O timer failed with %d.", errno);
+    RDB_MUTEX_UNLOCK_CHECK(m_reset_mutex);
+    return;
+  }
+
+  struct itimerspec timer_spec;
+  memset(&timer_spec, 0, sizeof(timer_spec));
+
+  // One time execution only for the watchdog. No interval.
+  timer_spec.it_value.tv_sec = m_write_timeout;
+
+  ret = timer_settime(m_io_check_watchdog_timer, 0, &timer_spec, nullptr);
+
+  if (unlikely(ret)) {
+    // NO_LINT_DEBUG
+    sql_print_warning("Setting time for a watchdog I/O timer failed with %d.",
+                      errno);
+    RDB_MUTEX_UNLOCK_CHECK(m_reset_mutex);
+    return;
+  }
+
+  m_io_in_progress.store(true);
+
+  // Verify the write access to all directories we care about.
+  for (const std::string &directory : m_dirs_to_check) {
+    ret = check_write_access(directory);
+
+    // We'll log a warning and attept to continue to see if the problem happens
+    // in other cases as well.
+    if (unlikely(ret != HA_EXIT_SUCCESS)) {
+      // NO_LINT_DEBUG
+      sql_print_warning("Unable to verify write access to %s (error code %d).",
+                        directory.c_str(), ret);
+    }
+  }
+
+  m_io_in_progress.store(false);
+
+  // Clean up the watchdog timer.
+  ret = timer_delete(m_io_check_watchdog_timer);
+
+  if (unlikely(ret)) {
+    // NO_LINT_DEBUG
+    sql_print_warning("Deleting the watchdog I/O timer failed with %d.", errno);
+  }
+
+  m_io_check_watchdog_timer = nullptr;
+
+  RDB_MUTEX_UNLOCK_CHECK(m_reset_mutex);
+}
+
+int Rdb_io_watchdog::check_write_access(const std::string &dirname) const {
+  DBUG_ASSERT(!dirname.empty());
+  DBUG_ASSERT(m_buf != nullptr);
+
+  const std::string fname = dirname + FN_DIRSEP + RDB_IO_DUMMY_FILE_NAME;
+
+  // O_DIRECT is a key flag here to make sure that we'll bypass the kernel's
+  // buffer cache.
+  int fd = open(fname.c_str(), O_WRONLY | O_DIRECT | O_CREAT | O_SYNC,
+                S_IRWXU | S_IWUSR);
+
+  if (unlikely(fd == -1)) {
+    return fd;
+  }
+
+  int ret = write(fd, m_buf, RDB_IO_WRITE_BUFFER_SIZE);
+
+  if (unlikely(ret != RDB_IO_WRITE_BUFFER_SIZE)) {
```

Comment:
I don't think so, if EINTR occurs, write is supposed to return -1, not 0.

Reply:
You are right re. EINTR. What's broken here is write returning success having written zero bytes. I don't know if it ever does that, but I'm unable to find why this would be a write contract violation. What's also broken - and, as opposed to zero bytes written with success case, actually happens - is writing more than zero but less than requested. Then the diagnostics one level up will be "Unable to verify write access to blah, errno 0"

Reply:
The absolutely proper fix is as usual wrapping write in an EINTR-, and returned bytes-checking loop in its own function, and every non-trivial program/lib should have one already. MySQL mysys has my_write for this, maybe MyRocks has another one too already?

Reply:
I was considering going the whole EINTR route for this similar to InnoDB and PerconaFT, but both implementations span layers and are quite messy for what amounts to a tiny, and fairly unimportant check in this unimportant class to see if there is write access to a file. MyRocks has no i/o layer wrapping because MyRocks itself, like TokuDB, does no i/o directly on its own. All i/o is done through the lower level dependent code (rocksdb, perconaft) and those do not expose their low level i/o APIs through their high level interfaces. So, anything and everything related to i/o or OS primitives that does not use the my_* has to be reinvented here at the storage engine layer. 

The only negative side effect that I can see in this thing from not fully and properly handling all cases where write does not write exactly the data expected is that we will have false sql_print_warnings done as a result.

I reiterate, I almost didn't even take this patch as I see it as a completely useless chunk of code to exist and run within the storage engine. It may be interesting as an experimental performance monitor or debugging tool, but useless in a 'soon to be production' engine. Just my $.02 worth

Reply:
I see sloppy code, I point it out. Up to you to decide where to cut losses.

Reply:
Ohh no doubt, and I totally appreciate it. I just wish that it didn't exist and need pointing out in the first place.

Reply:
Hmm, looks like it causes issues w/ OSX now too https://lists.launchpad.net/maria-developers/msg10851.html

Reply:
You know what, I really don't like what this patch is doing. My initial assertion that the only 'downside' is excessive sql_print_warnings, but that is not true, it aborts the server if there is an i/o timeout. I am going to remove the entirely. I do not think this behavior is at all appropriate and this is just not worth the effort to fix.

Reply:
Fine with me, just please update upstream with both the high level issues and the code changes so far. Re. macOS, it's OK not to care since it's not our supported platform, and someone with a Mac can always fix it later. Depending on what is the exact issue, the MariaDB fix is probably better reworked as a CMake check anyway.